staging/src/k8s.io/mount-utils/mount_helper_common.go

```diff
@@ -139,7 +139,7 @@ func removePathIfNotMountPoint(mountPath string, mounter Interface, extensiveMou
 // PathExists returns true if the specified path exists.
 // TODO: clean this up to use pkg/util/file/FileExists
 func PathExists(path string) (bool, error) {
-	_, err := os.Stat(path)
+	_, err := os.Lstat(path)
```

Comment:
does this fixes the issue? I thought lstat only ensures that if `path` is a symlink `lstat` will return information about `path` symlink rather than follow the symlink?

Reply:
I've tried this since I'm also not convinced this would have any impact and the issue #97031 is still reproducible. So I'd say no, this doesn't fix the issue.

Reply:
I tried to poke into this a bit and it seems like using `syscall.Access` instead of `os.Stat` might help.

Reply:
Forgive the drive-by comment, but since I had a repro I've tested the suggestions, and they all get "no such file or directory" on my system:

```
$ sudo ls /var/lib/kubelet/pods/823a42ab-1f77-4caa-b3ea-4e33fe43dd53/volume-subpaths/pv-smb/deployment-smb  
0
$ sudo ./stat /var/lib/kubelet/pods/5e50975f-6331-4e76-9968-09627864d30d/volume-subpaths/pv-smb/deployment-smb/0
checking /var/lib/kubelet/pods/5e50975f-6331-4e76-9968-09627864d30d/volume-subpaths/pv-smb/deployment-smb/0
os.Stat: stat /var/lib/kubelet/pods/5e50975f-6331-4e76-9968-09627864d30d/volume-subpaths/pv-smb/deployment-smb/0: no such file or directory
os.Lstat: lstat /var/lib/kubelet/pods/5e50975f-6331-4e76-9968-09627864d30d/volume-subpaths/pv-smb/deployment-smb/0: no such file or directory
syscall.Access: no such file or directory
```

from

```
        _, err := os.Stat(p)
        fmt.Printf("os.Stat: %v\n", err)
        _, err = os.Lstat(p)
        fmt.Printf("os.Lstat: %v\n", err)
        err = syscall.Access(p, syscall.F_OK)
        fmt.Printf("syscall.Access: %v\n", err)
```

Perhaps it would be better to call `umount` without trying to check for path existence? If the path doesn't exist after calling umount, it could suppress the error and print the expected "skipping" message, if that's the motivation for the check.

Reply:
@drigz in your case the first `ls` was run on a different directory than the subsequent test (the paths for `ls` and `./stat` test binary are different), so it's hard to say from the output you posted what really is wrong...

I actually tried just with a samba server without kubernetes involved to check the behaviour:
Mount the CIFS share:
```
[root@saiph smbtest]# mount -t cifs //server/share share
Password for root@//server/share: 
```
Create "subpath" directory in the samba share and bind mount it to a local "submount" mount point:
```
[root@saiph smbtest]# mkdir share/subpath
[root@saiph smbtest]# mkdir submount
[root@saiph smbtest]# mount -o bind share/subpath submount
[root@saiph smbtest]# ll
total 2032
drwxr-xr-x 2 root root       0 Nov 19 12:14 share
drwxr-xr-x 2 root root       0 Nov 19 12:14 submount
-rwxrwxr-x 1 root root 2079901 Nov 18 16:46 test
[root@saiph smbtest]# mount
<...>
//server/share on /root/smbtest/share type cifs (rw,relatime,vers=3.1.1,cache=strict,username=root,uid=0,noforceuid,gid=0,noforcegid,addr=192.168.1.1,file_mode=0755,dir_mode=0755,soft,nounix,serverino,mapposix,rsize=4194304,wsize=4194304,bsize=1048576,echo_interval=60,actimeo=1)
//server/share on /root/smbtest/submount type cifs (rw,relatime,vers=3.1.1,cache=strict,username=root,uid=0,noforceuid,gid=0,noforcegid,addr=192.168.1.1,file_mode=0755,dir_mode=0755,soft,nounix,serverino,mapposix,rsize=4194304,wsize=4194304,bsize=1048576,echo_interval=60,actimeo=1)
```
Remove the bind mounted dir from the original mount point and try to list the mountpoint:
```
[root@saiph smbtest]# rmdir share/subpath 
[root@saiph smbtest]# ll
ls: cannot access 'submount': No such file or directory
total 2032
drwxr-xr-x 2 root root       0 Nov 19 12:16 share
d????????? ? ?    ?          ?            ? submount
-rwxrwxr-x 1 root root 2079901 Nov 18 16:46 test
```
Note the strange output of `ls -l`. This is consistent with the comments in the issue #97031. Now -- `test` is a go binary calling `os.Stat` and then `syscall.Access` on the given path (very similarly to the code snippet from @drigz):
```
[root@saiph smbtest]# ./test submount 
Calling os.Stat(): Stat error: lstat submount: no such file or directory
Calling syscall.Access(): Access OK
```
Unmount of the bind mounted path also works:
```
[root@saiph smbtest]# umount submount 
[root@saiph smbtest]# mount
<...>
//server/share on /root/smbtest/share type cifs (rw,relatime,vers=3.1.1,cache=strict,username=root,uid=0,noforceuid,gid=0,noforcegid,addr=192.168.1.1,file_mode=0755,dir_mode=0755,soft,nounix,serverino,mapposix,rsize=4194304,wsize=4194304,bsize=1048576,echo_interval=60,actimeo=1)
[root@saiph smbtest]# ll
total 2036
drwxr-xr-x 2 root root       0 Nov 19 12:16 share
drwxrwxr-x 2 root root    4096 Nov 19 12:14 submount
-rwxrwxr-x 1 root root 2079901 Nov 18 16:46 test
```
... and also the `ls` output looks normal again.

I'm actually not sure what can break in case we remove the path check before the unmount, so I thought it would be safer to keep the current behaviour and just accomodate the code to this particular corner case.

Reply:
@tsmetana You are entirely correct: syscall.Access works fine for this on Linux, and I was testing the wrong path. Sorry for the noise!

Reply:
@drigz no problem! Thanks for testing it, actually. I think I might have a patch ready, just need to test it with the real thing and then I'll create a PR...

Reply:
...and no, won't be that easy: we also need to clean the subPath and that's where things start breaking in a slightly different way.

Reply:
There seems to be a similar poblem in `IsLikelyNotMountPoint` that gets called during cleanup: https://github.com/kubernetes/kubernetes/blob/aff056d8a197f6a404ad5e02210ca662d16c3dbe/staging/src/k8s.io/mount-utils/mount_linux.go#L337-L352

Unfortunatelly the `os.Stat` result is being used there...